timemachine/fe/free_energy.py

```diff
@@ -1120,6 +1136,91 @@ def compute_dense():
     return U_kl
 
 
+def make_u_kl_fxn(trajs, initial_states):
+    """fxn(k, l) = "trajs[k] evaluated in ensembles[l]"
+
+    usage note: be careful of axis-ordering convention, see: https://github.com/proteneer/timemachine/issues/1100
+    """
+
+    summed_potentials = [make_summed_potential(s.potentials) for s in initial_states]
+
+    # validate assumption that initial states all have compatible potentials / ensembles
+    kBTs = [BOLTZ * state.integrator.temperature for state in initial_states]
+    s_0 = initial_states[0]
+    for s in initial_states[1:]:
+        assert_ensembles_compatible(s_0, s)
+        assert_potentials_compatible(s_0.potentials, s.potentials)
+
+    sp = summed_potentials[0]
+    sp_gpu = sp.potential.to_gpu(np.float32)
+
+    def batch_U_fxn(xs, ps, bs, x_idxs, p_idxs):
+        Us = sp_gpu.unbound_impl.execute_batch_sparse(xs, ps, bs, x_idxs, p_idxs, False, False, True)[2]
+        return np.nan_to_num(Us, nan=+np.inf)
+
+    def u_kl(k, l):
+        coords = trajs[k].frames
+        boxes = trajs[k].boxes
+
+        params = np.array([summed_potentials[l].params])
+
+        coords_batch_idxs = np.arange(len(coords)).astype(np.uint32)
+        params_batch_idxs = np.zeros_like(coords_batch_idxs).astype(np.uint32)
+
+        Us = batch_U_fxn(coords, params, boxes, coords_batch_idxs, params_batch_idxs)
```

Comment:
nit: this doesn't seem like it's using the full generality of `execute_batch_sparse` -- would `execute_batch` suffice in the definition of `batch_U_fxn` for clarity?

Reply:
Either `execute_batch` or `execute_batch_sparse` I think would suffice, but the `execute_batch_sparse` interface is slightly more explicit

Reply:
OK, this is only a minor stylistic preference (i.e. using the least powerful abstraction available), but can see either argument, so not a blocker.